src/main/java/com/uber/cadence/internal/sync/WorkflowThreadImpl.java

```diff
@@ -220,18 +224,31 @@ public boolean isCancelRequested() {
 
   @Override
   public void start() {
+    int attempts = 2;
+    for (int i = 0; i < attempts; i++) {
+      try {
+        attemptStart();
+        break;
+      } catch (RejectedExecutionException e) {
+        if (i == attempts - 1 || cache == null) {
+          throw new Error(
+              "Not enough threads to execute workflows. "
+                  + "If this message appears consistently either WorkerOptions.maxConcurrentWorklfowExecutionSize "
+                  + "should be decreased or WorkerOptions.maxWorkflowThreads increased.");
+        }
+        // Reset state
+        context.setStatus(Status.CREATED);
+        cache.evictNext();
```

Comment:
Add metric for the thread forced evictions. And for all other cache related metrics if not added yet. It can be done in a separate PR.

Reply:
Doing that in a separate PR